tests/phpunit/tests/base-sender.php

```diff
@@ -3,74 +3,67 @@
 class Test_WPNotify_BaseSender extends WPNotify_TestCase {
 
 	/**
-	 * @param string|array $sender
-	 * @param string       $expected_json
+	 * @param array  $sender_params
+	 * @param string $expected_json
 	 *
 	 * @dataProvider data_provider_senders
 	 */
-	public function test_it_can_be_json_encoded( $sender, $expected_json ) {
+	public function test_it_can_be_json_encoded( $sender_params, $expected_json ) {
 
-		$sender_instance = null;
+		$sender_reflection = new ReflectionClass( 'WPNotify_BaseSender' );
+		$sender            = $sender_reflection->newInstanceArgs( array_values( $sender_params ) );
 
-		if ( is_string( $sender ) ) {
-			$sender_instance = new WPNotify_BaseSender( $sender );
-		} else if ( is_array( $sender ) ) {
-			$sender_instance = new WPNotify_BaseSender( $sender['name'], $sender['image'] );
-		}
-
-		$sender_encoded = json_encode( $sender_instance );
+		$sender_encoded = json_encode( $sender );
 
 		$this->assertEquals( $expected_json, $sender_encoded );
 	}
 
 	/**
-	 * @param string|array $sender
-	 * @param string       $json
+	 * @param array  $sender_params
+	 * @param string $json
 	 *
 	 * @dataProvider data_provider_senders
 	 */
-	public function test_it_can_be_instantiated_from_json( $sender, $json ) {
+	public function test_it_can_be_instantiated_from_json( $sender_params, $json ) {
 
 		$testee = WPNotify_BaseSender::json_unserialize( $json );
 
-		$sender_name     = '';
-		$sender_instance = null;
-
-		if ( is_string( $sender ) ) {
-			$sender_name     = $sender;
-			$sender_instance = new WPNotify_BaseSender( $sender );
-		} else if ( is_array( $sender ) ) {
-			$sender_name     = $sender['name'];
-			$sender_instance = new WPNotify_BaseSender( $sender['name'], $sender['image'] );
-		}
+		$sender_reflection = new ReflectionClass( 'WPNotify_BaseSender' );
+		/** @var WPNotify_BaseSender $sender */
+		$sender = $sender_reflection->newInstanceArgs( array_values( (array) $sender_params ) );
 
 		$this->assertInstanceOf( 'WPNotify_BaseSender', $testee );
-		$this->assertEquals( $sender_name, $testee->get_name() );
+		$this->assertEquals( $sender->get_name(), $testee->get_name() );
+
+		$this->assertEquals( $sender->get_image(), $testee->get_image() );
 
 		/**
 		 * If an Image has been sent
 		 */
 		if ( $testee->get_image() ) {
-			$this->assertInstanceOf( 'WPNotify_BaseImage', $testee->get_image() );
-			$this->assertEquals( $sender_instance->get_image()->get_source(), $testee->get_image()->get_source() );
-			$this->assertEquals( $sender_instance->get_image()->get_alt(), $testee->get_image()->get_alt() );
+			$this->assertInstanceOf( get_class( $sender->get_image() ), $testee->get_image() );
+			$this->assertEquals( $sender->get_image()->get_source(), $testee->get_image()->get_source() );
+			$this->assertEquals( $sender->get_image()->get_alt(), $testee->get_image()->get_alt() );
```

Comment:
If line 38 passes, is there any way for lines 45 to 46 to fail?

I mean if `$sender->get_image() === $testee->get_image()`, then their classes, `source` property and `alt` property probably match as well.

But  `$sender->get_image() === $testee->get_image()` imply that the two `image` properties refer to the same instance... Is this something we intended here?

Reply:
PHPUnit does not perform a deep comparison between the two `WPNotify_BaseImage` objects, so we need to compare their individual properties.